app/models/manageiq/providers/openstack/inventory/parser/cloud_manager.rb

```diff
@@ -0,0 +1,458 @@
+      o[:ems_ref] = stack.id.to_s
+      o[:name] = stack.stack_name
+      o[:description] = stack.description
+      o[:status] = stack.stack_status
+      o[:status_reason] = stack.stack_status_reason
+      o[:parent] = persister.orchestration_stacks.lazy_find(stack.parent)
+      o[:resources] = orchestration_stack_resources(stack)
+      o[:outputs] = orchestration_stack_outputs(stack)
+      o[:parameters] = orchestration_stack_parameters(stack)
+      o[:orchestration_template] = orchestration_template(stack)
+      o[:cloud_tenant] = persister.cloud_tenants.lazy_find(stack.service.current_tenant["id"])
+    end
+  end
+
+  def vms
+    related_infra_ems = collector.manager.provider.try(:infra_ems)
+    hosts = related_infra_ems.try(:hosts)
+
+    collector.servers.each do |s|
+      if hosts
+        # Find associated host from OpenstackInfra
+        filtered_hosts = hosts.select do |h|
+          h.hypervisor_hostname && s.os_ext_srv_attr_host && s.os_ext_srv_attr_host.include?(h.hypervisor_hostname.downcase)
+        end
+        parent_host = filtered_hosts.first
+        parent_cluster = parent_host.try(:ems_cluster)
+      else
+        parent_host = nil
+        parent_cluster = nil
+      end
+
+      availability_zone = s.availability_zone.blank? ? "null_az" : s.availability_zone
+
+      server = persister.vms.find_or_build(s.id.to_s)
+      server[:type] = "ManageIQ::Providers::Openstack::CloudManager::Vm"
+      server[:uid_ems] = s.id
+      server[:ems_ref] = s.id
+      server[:name] = s.name
+      server[:vendor] = "openstack"
+      server[:raw_power_state] = s.state || "UNKNOWN"
+      server[:connection_state] = "connected"
+      server[:location] = "unknown"
+      server[:host] = parent_host
+      server[:ems_cluster] = parent_cluster
+      server[:availability_zone] = persister.availability_zones.lazy_find(availability_zone)
+      server[:key_pairs] = [persister.key_pairs.lazy_find(s.key_name)].compact
+      server[:cloud_tenant] = persister.cloud_tenants.lazy_find(s.tenant_id.to_s)
+      server[:parent] = persister.miq_templates.lazy_find(s.image["id"]) unless s.image["id"].nil?
+
+      # to populate the hardware, we need some fields from the flavor object
+      # that we don't already have from the flavor field on the server details
+      # returned from the openstack api. It's possible that no such flavor was found
+      # due to some intermittent network issue or etc, so we use try to not break.
+      flavor = collector.flavors_by_id[s.flavor["id"].to_s]
+      if flavor.nil?
+        # the flavor might be private, which the flavor list doesn't seem to handle
+        # correctly. fetch it now and build an inventory object for it.
+        flavor = collector.private_flavor(s.flavor["id"].to_s)
+        make_flavor(flavor) unless flavor.nil?
+      end
+      server[:flavor] = persister.flavors.lazy_find(s.flavor["id"].to_s)
+      server[:hardware] = persister.hardwares.find_or_build(s.id)
```

Comment:
this doesn't need to be assigned to server[:hardware], again, it can be creating an unbreakble cycle. This relation is solved by server[:hardware][:vm_or_template] = persister.vms.lazy_find(s.id)

So by default, we should be saving only what's in the DB table. And in some cases, it can be beneficial to store into relation. But it can be problematic to store in the relation and having this object as part of the unique key on the other side.